frame/elections-phragmen/src/lib.rs

```diff
@@ -346,42 +421,67 @@ decl_module! {
 			origin,
 			votes: Vec<T::AccountId>,
 			#[compact] value: BalanceOf<T>,
-		) {
+		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
 
+			// votes should not be empty and more than `MAXIMUM_VOTE` in any case.
 			ensure!(votes.len() <= MAXIMUM_VOTE, Error::<T>::MaximumVotesExceeded);
 			ensure!(!votes.is_empty(), Error::<T>::NoVotes);
 
 			let candidates_count = <Candidates<T>>::decode_len().unwrap_or(0);
 			let members_count = <Members<T>>::decode_len().unwrap_or(0);
 			let runners_up_count = <RunnersUp<T>>::decode_len().unwrap_or(0);
 
+			// can never submit a vote of there are no members, and cannot submit more votes than
+			// all potential vote targets.
 			// addition is valid: candidates, members and runners-up will never overlap.
 			let allowed_votes = candidates_count + members_count + runners_up_count;
-
 			ensure!(!allowed_votes.is_zero(), Error::<T>::UnableToVote);
 			ensure!(votes.len() <= allowed_votes, Error::<T>::TooManyVotes);
 
 			ensure!(value > T::Currency::minimum_balance(), Error::<T>::LowBalance);
 
-			// first time voter. Reserve bond.
-			if !Self::is_voter(&who) {
-				T::Currency::reserve(&who, T::VotingBond::get())
+			// Reserve bond.
+			let new_deposit = Self::deposit_of(votes.len());
+			let Voter { votes: old_votes, deposit: old_deposit, .. } = <Voting<T>>::get(&who);
+			let maybe_refund = if old_votes.is_empty() {
+				// First time voter. Get full deposit.
```

Comment:
I usually prefer not to admit that old_votes.is_empty is same as first time voter, because maybe in the future some mecanism can remove some voter or something and this property won't be respected anymore.
* why not using the fact that old_deposit is 0 ?
* or use Voting as `Option<Voter>`
* or just remove this special case which just saves almost nothing compare to the case `new_deposit > old_deposit`

Reply:
I like the third option the best 👍

Reply:
Addressed.